### PR TITLE
fix notebook image build

### PR DESCRIPTION
**Pull request by lluunn**
With the change, I can build the notebook image.
The new one is v20191216-v0.7.0-rc.5-61-g2da14890-dirty-d0bbf2, will update the manifest.

/cc @kimwnasptd @fediazgon @jlewi 
related to https://github.com/kubeflow/kubeflow/issues/4582

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4583)

### CODE REVIEW

**Comment by fediazgon** [comment]
Just tried locally and works fine for me.

File: components/notebook-controller/go.mod [hunk 1]
```diff
@@ -4,19 +4,11 @@ go 1.12
 
 require (
 	github.com/go-logr/logr v0.1.0
-	github.com/gogo/protobuf v1.2.1 // indirect
-	github.com/json-iterator/go v1.1.6 // indirect
-	github.com/modern-go/reflect2 v1.0.1 // indirect
-	github.com/onsi/ginkgo v1.8.0 // indirect
-	github.com/onsi/gomega v1.5.0 // indirect
+	github.com/kubeflow/kubeflow v0.7.0-rc.5.0.20191205174032-2da148904248
```
Comment (jlewi) [inline components/notebook-controller/go.mod]: Is this pinning to a specific version of kubeflow/kubeflow? Do we want to do that? 
Why is github.com/kubeflow/kubeflow listed go.mod? Isn't github.com/kubeflow/kubeflow the module we are building?

File: components/notebook-controller/go.sum [hunk 2]
```diff
@@ -38,6 +44,13 @@ github.com/kisielk/gotool v1.0.0/go.mod h1:XhKaO+MFFWcvkIS/tQcRk01m1F5IRFswLeQ+o
 github.com/kr/pretty v0.1.0/go.mod h1:dAy3ld7l9f0ibDNOQOHHMYYIIbhfbHSm3C4ZsoJORNo=
 github.com/kr/pty v1.1.1/go.mod h1:pFQYn66WHrOpPYNljwOMqo10TkYh1fy3cYio2l3bCsQ=
 github.com/kr/text v0.1.0/go.mod h1:4Jbv+DJW3UT/LiOwJeYQe1efqtUx/iVham/4vfdArNI=
+github.com/kubeflow/kubeflow v0.7.0-rc.5.0.20191205174032-2da148904248 h1:VrZRKsJuXQ9Wsxmk6Sjcj+ZB74g4FchomlJyHTdCfWk=
```
Comment (jlewi) [inline components/notebook-controller/go.sum]: Same question as above

**Comment by jlewi** [comment]
@lluunn Could you explain what the problem is and how this fixes it? Do you know why this wasn't caught by tests? Is our postsubmit and periodic tests red?

**Comment by jlewi** [comment]
@kkasravi could you please take a look at the use of go.mod here and provide some guidance.

It looks like we have multiple modules in kubeflow/kubeflow (e.g. notebook-controller) is one that depends on other modules (e.g. common/reconcile-helper).

When building notebook-controller we'd like it to depend on the other modules checked out in the same repo.

Does that seem right? And if so how do we do that?

**Comment by lluunn** [comment]
The problem is if you try `make build-gcr` from master, it doesn't work. I think presubmit doesn't build the image.
The error was that it cannot load https://github.com/kubeflow/kubeflow/tree/master/components/common/reconcilehelper, which is moved there from notebook controller in https://github.com/kubeflow/kubeflow/pull/4312.

I tried the suggestion in https://github.com/kubeflow/kubeflow/pull/4214#issuecomment-566215307 to do `go mod edit -require github.com/kubeflow/kubeflow@master`. Not sure if it's the best way forward.

**Comment by jlewi** [comment]
@lluunn could you open up a bug about the jupyter E2E test? I would expect that to build the image from head and test it.

It looks to me though like the way you are fixing the bug is pinning to a specific commit of kubeflow/kubeflow to find reconcilehelper

```
require (
  github.com/kubeflow/kubeflow v0.7.0-rc.5.0.20191205174032-2da148904248
)
```

I would expect that means we don't pick up any local changes to kubeflow/kubeflow when building the image. Is that what we want? e.g. if we were to end up making a fix to reconcilehelper would the fix get picked up?

I guess a smoke test would be to add a syntax error in reconcilehelper and see if it still builds. If it does then that would confirm you aren't picking up locally modified code.

**Comment by swiftdiaries** [comment]
One possible approach:
We could move the [reconcilehelper util](https://github.com/kubeflow/kubeflow/blob/master/components/common/reconcilehelper/util.go) into either the notebook-controller or the tensorboard-controller and have the other one share the implementation.

**Comment by lluunn** [comment]
I moved back the reconcile util.

**Comment by jlewi** [comment]
If we think moving the reconcileHelper into the notebook package is the best solution then we can go with that.

However, if IIUC we should be able to make the original package layout work just by setting up go.mod correctly. 

See the second stackoverflow answer here
https://stackoverflow.com/questions/52026284/accessing-local-packages-within-a-go-module-go-1-11

Do we just need to add a replace directive to point the module at a relative path?

See also this doc
https://github.com/golang/go/wiki/Modules#when-should-i-use-the-replace-directive

One of the main use cases for the replace directive is a multi-module project and it looks like that might be what we have here.

**Comment by jlewi** [comment]
Closing in favor of kubeflow/kubeflow#4631 which updates go.mod

**Pull request closed without merging**